proto/internal/temporal/server/api/persistence/v1/executions.proto

```diff
@@ -249,6 +255,41 @@ message WorkflowExecutionInfo {
     VersionedTransition signal_request_ids_last_update_versioned_transition = 94;
 
     repeated StateMachineTombstoneBatch sub_state_machine_tombstone_batches = 95;
+
+    // Holds all the information about versioning for this particular execution.
+    VersioningInfo versioning_info = 96;
+
+    message VersioningInfo {
+        // Determines how server should treat this execution when workers are upgraded.
+        // When present it means that this workflow is versioned. This is set after an execution
+        // completes its first workflow task on a versioned worker.
+        temporal.api.enums.v1.VersioningBehavior behavior = 1;
+        // The worker deployment to which this workflow is assigned currently.
+        // Must be present if and only if `behavior` is set.
+        temporal.api.common.v1.WorkerDeployment deployment = 2;
+        // Manual override for execution's versioning behavior. Takes precedence over `behavior`.
+        temporal.api.enums.v1.VersioningBehavior behavior_override = 3;
+        // Used to manually pin the execution to a deployment. Must be set if and only if
+        // `behavior_override` is PINNED. Takes precedence over `deployment`.
+        temporal.api.common.v1.WorkerDeployment deployment_override = 4;
+        // When present, indicates the workflow is being redirected to a different deployment.
+        // A redirect can only exist during the lifetime of a pending workflow task.
+        // If the pending workflow task completes (at the next WorkflowTaskCompleted event), the
+        // redirect is considered complete and the workflow's deployment is updated. If the pending
+        // workflow task fails or times out, then the redirect is canceled and workflow remains on
+        // the previous deployment.
```

Comment:
why do we want to cancel the redirect if the pending WFT fails or times out? what if it timed out due to reasons totally unrelated to versioning? 

is the idea that it's ok to cancel the redirect because if the WF really does need to change deployments or change behavior, and the timeout was due to a transient issue, the next WFT will re-initiate the cancelled redirect, it will succeed, and the redirect will complete?

Reply:
yes, that's the thinking. basically, if its a safe move redirect we want to fail it. If it's from matching, it'll repeat in the next task redirect.